File: app/vocab.py
"""Per-tenant-type label vocabulary.

The underlying model is identical across vendor types — there's always a token,
a slot, a queue, an "in-consult" person, a "done" state. Only the user-facing
labels differ between a clinic and a salon. This module is the single source of
truth for that.

Templates pull labels via the `vocab` global (a dict keyed by `tenant_type` or
an explicit override). Falls back to the `clinic` mapping when a key is missing
or the tenant_type is unknown.
"""
from __future__ import annotations

from app.models import TenantType

# Vocab keys consumed by templates. Add a new key here when you need it.
_DEFAULT = {
    "workspace":         "clinic",          # what is the business called?
    "workspace_titled":  "Clinic",
    "entity_singular":   "patient",         # who comes in for service?
    "entity_plural":     "patients",
    "entity_titled":     "Patient",
    "provider":          "doctor",          # who provides the service?
    "provider_titled":   "Doctor",
    "session":           "consult",         # what's a single encounter called?
    "session_titled":    "Consult",
    "session_progress":  "in consult",
    "staff":             "receptionist",
    "staff_titled":      "Receptionist",
    "reason_label":      "Reason for visit",
}


VOCABS: dict[TenantType, dict[str, str]] = {
    TenantType.clinic: {
        **_DEFAULT,
    },
    TenantType.dental: {
        **_DEFAULT,
        "workspace": "practice", "workspace_titled": "Practice",
        "provider": "dentist", "provider_titled": "Dentist",
        "session": "appointment", "session_titled": "Appointment",
        "session_progress": "in chair",
        "reason_label": "Procedure",
    },
    TenantType.salon: {
        **_DEFAULT,
        "workspace": "salon", "workspace_titled": "Salon",
        "entity_singular": "customer", "entity_plural": "customers", "entity_titled": "Customer",
        "provider": "stylist", "provider_titled": "Stylist",
        "session": "service", "session_titled": "Service",
        "session_progress": "in chair",
        "staff": "front desk", "staff_titled": "Front desk",
        "reason_label": "Service requested",
    },
    TenantType.spa: {
        **_DEFAULT,
        "workspace": "spa", "workspace_titled": "Spa",
        "entity_singular": "guest", "entity_plural": "guests", "entity_titled": "Guest",
        "provider": "therapist", "provider_titled": "Therapist",
        "session": "session", "session_titled": "Session",
        "session_progress": "in session",
        "staff": "front desk", "staff_titled": "Front desk",
        "reason_label": "Service requested",
    },
    TenantType.vet: {
        **_DEFAULT,
        "workspace": "clinic", "workspace_titled": "Clinic",
        "entity_singular": "pet", "entity_plural": "pets", "entity_titled": "Pet",
        "provider": "vet", "provider_titled": "Vet",
        "session": "visit", "session_titled": "Visit",
        "session_progress": "in exam",
        "reason_label": "Reason for visit",
    },
    TenantType.other: {
        **_DEFAULT,
        "workspace": "business", "workspace_titled": "Business",
        "entity_singular": "customer", "entity_plural": "customers", "entity_titled": "Customer",
        "provider": "owner", "provider_titled": "Owner",
        "session": "appointment", "session_titled": "Appointment",
        "session_progress": "with provider",
        "staff": "staff", "staff_titled": "Staff",
        "reason_label": "Notes",
    },
}


def vocab_for(tenant_type: TenantType | str | None) -> dict[str, str]:
    if tenant_type is None:
        return VOCABS[TenantType.clinic]
    if isinstance(tenant_type, str):
        try:
            tenant_type = TenantType(tenant_type)
        except ValueError:
            return VOCABS[TenantType.clinic]
    return VOCABS.get(tenant_type, VOCABS[TenantType.clinic])


# Labels shown on the public signup page when picking a vendor type
TENANT_TYPE_DISPLAY: list[tuple[TenantType, str, str]] = [
    (TenantType.clinic,  "Clinic",      "Homeopathy, GP, paediatrics, physiotherapy, etc."),
    (TenantType.dental,  "Dental",      "Dentist, orthodontist."),
    (TenantType.salon,   "Salon",       "Hair, beauty, nails, barbershop."),
    (TenantType.spa,     "Spa",         "Massage, wellness, day spa."),
    (TenantType.vet,     "Vet",         "Veterinary clinic."),
    (TenantType.other,   "Other",       "Any appointment-based service business."),
]

File: frontend/lib/vocab.ts
// Per-tenant-type label vocabulary — port of app/vocab.py.
export type TenantTypeKey =
  | "clinic"
  | "salon"
  | "spa"
  | "dental"
  | "vet"
  | "other";

export type Vocab = {
  workspace: string;
  workspaceTitled: string;
  entitySingular: string;
  entityPlural: string;
  entityTitled: string;
  provider: string;
  providerTitled: string;
  session: string;
  sessionTitled: string;
  sessionProgress: string;
  staff: string;
  staffTitled: string;
  reasonLabel: string;
};

const DEFAULT: Vocab = {
  workspace: "clinic",
  workspaceTitled: "Clinic",
  entitySingular: "patient",
  entityPlural: "patients",
  entityTitled: "Patient",
  provider: "doctor",
  providerTitled: "Doctor",
  session: "consult",
  sessionTitled: "Consult",
  sessionProgress: "in consult",
  staff: "receptionist",
  staffTitled: "Receptionist",
  reasonLabel: "Reason for visit",
};

export const VOCABS: Record<TenantTypeKey, Vocab> = {
  clinic: { ...DEFAULT },
  dental: {
    ...DEFAULT,
    workspace: "practice",
    workspaceTitled: "Practice",
    provider: "dentist",
    providerTitled: "Dentist",
    session: "appointment",
    sessionTitled: "Appointment",
    sessionProgress: "in chair",
    reasonLabel: "Procedure",
  },
  salon: {
    ...DEFAULT,
    workspace: "salon",
    workspaceTitled: "Salon",
    entitySingular: "customer",
    entityPlural: "customers",
    entityTitled: "Customer",
    provider: "stylist",
    providerTitled: "Stylist",
    session: "service",
    sessionTitled: "Service",
    sessionProgress: "in chair",
    staff: "front desk",
    staffTitled: "Front desk",
    reasonLabel: "Service requested",
  },
  spa: {
    ...DEFAULT,
    workspace: "spa",
    workspaceTitled: "Spa",
    entitySingular: "guest",
    entityPlural: "guests",
    entityTitled: "Guest",
    provider: "therapist",
    providerTitled: "Therapist",
    session: "session",
    sessionTitled: "Session",
    sessionProgress: "in session",
    staff: "front desk",
    staffTitled: "Front desk",
    reasonLabel: "Service requested",
  },
  vet: {
    ...DEFAULT,
    entitySingular: "pet",
    entityPlural: "pets",
    entityTitled: "Pet",
    provider: "vet",
    providerTitled: "Vet",
    session: "visit",
    sessionTitled: "Visit",
    // Vet patients are examined like clinic patients — "in consult" reads
    // naturally for both. "in exam" was overly clinical and didn't match
    // how clinics talk about a doctor seeing a pet.
    sessionProgress: "in consult",
  },
  other: {
    ...DEFAULT,
    workspace: "business",
    workspaceTitled: "Business",
    entitySingular: "customer",
    entityPlural: "customers",
    entityTitled: "Customer",
    provider: "owner",
    providerTitled: "Owner",
    session: "appointment",
    sessionTitled: "Appointment",
    // "with provider" was a placeholder. "in session" is generic and
    // recognisable across coaching, legal, notary, and services at large.
    sessionProgress: "in session",
    staff: "staff",
    staffTitled: "Staff",
    reasonLabel: "Notes",
  },
};

export function vocabFor(tenantType: string | null | undefined): Vocab {
  if (!tenantType) return VOCABS.clinic;
  if (tenantType in VOCABS) return VOCABS[tenantType as TenantTypeKey];
  return VOCABS.clinic;
}

// The subset the mobile app needs — derived from the same map so it
// can't drift. Served alongside each clinic detail response.
export type MobileVocab = {
  sessionLabel: string; // upper-case pill under clinic name in cards
  bookCta: string;      // button text on the clinic detail screen
  tenantLabel: string;  // "Clinic" / "Salon" / "Dental" row on cards
};

const MOBILE_BOOK_CTA: Record<TenantTypeKey, string> = {
  clinic: "Book a consult",
  dental: "Book a checkup",
  salon: "Book a chair",
  spa: "Book a session",
  vet: "Book a visit",
  other: "Book a slot",
};

const MOBILE_TENANT_LABEL: Record<TenantTypeKey, string> = {
  clinic: "Clinic",
  dental: "Dental",
  salon: "Salon",
  spa: "Spa",
  vet: "Vet",
  other: "Business",
};

export function mobileVocabFor(tenantType: string | null | undefined): MobileVocab {
  const key: TenantTypeKey =
    tenantType && tenantType in VOCABS ? (tenantType as TenantTypeKey) : "clinic";
  const v = VOCABS[key];
  return {
    // v.session is lower-case ("consult", "session", …); mobile card
    // renders it uppercased.
    sessionLabel: v.session.toUpperCase(),
    bookCta: MOBILE_BOOK_CTA[key],
    tenantLabel: MOBILE_TENANT_LABEL[key],
  };
}
